{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "6de49d5e-d725-45f6-92e1-6831e18d8891",
   "metadata": {},
   "source": [
    "# Load libraries and Themes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "35d7dba7-aec1-4e24-9702-d0eb6d41a8e3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "suppressPackageStartupMessages({\n",
    "    suppressWarnings({\n",
    "        library(Seurat)\n",
    "        library(ggplot2)\n",
    "        library(tidyverse)\n",
    "        library(presto)\n",
    "        library(SeuratDisk)\n",
    "        library(DESeq2)\n",
    "        library(edgeR)\n",
    "        library(EnhancedVolcano)\n",
    "        library(Rsamtools)\n",
    "        library(svglite)\n",
    "        library(viridis)\n",
    "        library(pals)\n",
    "        library(scales)\n",
    "        library(sva)\n",
    "        })})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39ea3e8b-4e9f-4694-a0dc-58f4c95dd9af",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Color Palettes\n",
    "    \n",
    "palette.treatment <- c(\n",
    "    \"#AF0000\", # ALDO\n",
    "    \"#C1C1C1\", # CTRL\n",
    "    \"#007D7D\"  # REC\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "594f95d6-1f38-411c-aed5-ad8b70d5a04f",
   "metadata": {},
   "outputs": [],
   "source": [
    "umap_theme <- theme(\n",
    "  axis.line=element_blank(),\n",
    "  axis.text.x=element_blank(),\n",
    "  axis.text.y=element_blank(),\n",
    "  axis.ticks=element_blank(),\n",
    "  axis.title.x=element_blank(),\n",
    "  axis.title.y=element_blank(),\n",
    "  panel.background=element_blank(),\n",
    "  panel.border=element_blank(),\n",
    "  panel.grid.major=element_blank(),\n",
    "  panel.grid.minor=element_blank()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c5dd502-239e-4284-a041-5e787e6f7d0f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "setwd(\"/media/daten/dmeral/scseq_analysis/2024_LV_CTRL_ALDO_REC\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1051729-5b26-4121-9bfd-e0d177aea70f",
   "metadata": {},
   "outputs": [],
   "source": [
    "set.seed(1234)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10c1839a-42ba-4634-b6e5-4a947254709c",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "obj <- LoadH5Seurat(\"seurat_objects/obj_seu_merge_harmony_sgl_addmodule_rename_CMcomb_onlyprotcod_ccscore.h5seurat\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1a694d3-a03d-4f22-b462-4ab98e674aa1",
   "metadata": {},
   "source": [
    "# DEG analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "651d469e-35ec-4967-bcee-0ffc4f1b1cb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_annotations <- list(\n",
    "    \"0\" = \"CM_0\",\n",
    "    \"1\" = \"EC-cap\",\n",
    "    \"2\" = \"FB\",\n",
    "    \"3\" = \"PER\",\n",
    "    \"4\" = \"MΦ\",\n",
    "    \"5\" = \"EC-art\",\n",
    "    \"6\" = \"EC-end\",\n",
    "    \"7\" = \"SMC\",\n",
    "    \"8\" = \"CM_1\",\n",
    "    \"9\" = \"EC-lym\",\n",
    "    \"10\" = \"CYC\",\n",
    "    \"11\" = \"CM_2\",\n",
    "    \"12\" = \"BC\",\n",
    "    \"13\" = \"SC\",\n",
    "    \"14\" = \"TC\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff60d586-3252-426b-946a-700d37c1c82e",
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_annotations_CMcomb <- list(\n",
    "    \"0\" = \"CM\",\n",
    "    \"1\" = \"EC-cap\",\n",
    "    \"2\" = \"FB\",\n",
    "    \"3\" = \"PER\",\n",
    "    \"4\" = \"MΦ\",\n",
    "    \"5\" = \"EC-art\",\n",
    "    \"6\" = \"EC-end\",\n",
    "    \"7\" = \"SMC\",\n",
    "    \"8\" = \"CM\",\n",
    "    \"9\" = \"EC-lym\",\n",
    "    \"10\" = \"CYC\",\n",
    "    \"11\" = \"CM\",\n",
    "    \"12\" = \"BC\",\n",
    "    \"13\" = \"SC\",\n",
    "    \"14\" = \"TC\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b4fe8a9-05fc-426e-b48b-9c5e3c789fd2",
   "metadata": {},
   "source": [
    "## Wilcoxon Rank Sum test all combinations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea062fcd-741e-4fe6-9abd-13ddb7da77cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "doAnalysis <- function(LA_LV, cluster_annotations) {\n",
    "    # Subset based on chamber\n",
    "    seurat_table <- subset(x = obj, subset = chamber == LA_LV)\n",
    "   \n",
    "    # Set treatment as identity class\n",
    "    Idents(seurat_table) <- seurat_table@meta.data$\"treatment\"\n",
    "    \n",
    "    # Initialize a list to store DE results for each comparison\n",
    "    results <- list()\n",
    "    \n",
    "    # Loop over each cluster annotation\n",
    "    for (cl in names(cluster_annotations)) {\n",
    "        name_of_clust <- cluster_annotations[[cl]]\n",
    "        \n",
    "        # Subset by cluster\n",
    "        my_subseur <- subset(x = seurat_table, subset = seurat_clusters == cl)\n",
    "        my_subseur <- NormalizeData(my_subseur)\n",
    "        my_subseur <- ScaleData(my_subseur)\n",
    "        \n",
    "        # Perform comparisons for each pairwise condition\n",
    "        comparisons <- list(\n",
    "            \"ALDO_vs_CTRL\" = c(\"ALDO\", \"CTRL\"),\n",
    "            \"REC_vs_CTRL\" = c(\"REC\", \"CTRL\"),\n",
    "            \"ALDO_vs_REC\" = c(\"ALDO\", \"REC\")\n",
    "        )\n",
    "        \n",
    "        # Loop through each comparison\n",
    "        cluster_comparison_results <- list()\n",
    "        for (comparison_name in names(comparisons)) {\n",
    "            ident.1 <- comparisons[[comparison_name]][1]\n",
    "            ident.2 <- comparisons[[comparison_name]][2]\n",
    "            \n",
    "            # Find markers\n",
    "            my_markers <- FindMarkers(my_subseur, ident.1 = ident.1, ident.2 = ident.2, assay = \"RNA\", test.use = \"wilcox\", logfc.threshold = 0) # modify treshold\n",
    "            \n",
    "            # Save to CSV\n",
    "            file_name <- paste0(\"DEGs/nothreshold/\", comparison_name, \"_WRSum/degs_wilcox_\", LA_LV, \"_\", name_of_clust, \".csv\") # notreshold for logfc.threshold = 0\n",
    "            write.csv(my_markers, file = file_name)\n",
    "            \n",
    "            # Store results\n",
    "            cluster_comparison_results[[comparison_name]] <- my_markers\n",
    "        }\n",
    "        \n",
    "        # Store the results for this cluster\n",
    "        results[[name_of_clust]] <- cluster_comparison_results\n",
    "    }\n",
    "    \n",
    "    return(results)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c2fbb68-4dc5-48e2-9400-1d83885c31df",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "doAnalysis(\"LV\", cluster_annotations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "017b2049-31a0-4031-ae9d-4ec498ad56d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# For combined CM cluster\n",
    "doCMAnalysis <- function(LA_LV) {\n",
    "    # Subset the Seurat object for the selected chamber\n",
    "    seurat_table <- subset(obj, subset = chamber == LA_LV)\n",
    "    \n",
    "    # Assign \"treatment\" as the active identity class\n",
    "    Idents(seurat_table) <- seurat_table@meta.data$\"treatment\"\n",
    "    \n",
    "    # Subset for the combined \"CM\" cluster (using Cardiomyocyte column)\n",
    "    my_subseur <- subset(seurat_table, subset = Cardiomyocyte == \"CM\")\n",
    "    \n",
    "    # Check if there are any cells in the CM cluster\n",
    "    if (ncol(my_subseur) == 0) {\n",
    "        stop(\"No cells found in CM cluster.\")\n",
    "    }\n",
    "    \n",
    "    # Preprocess data for the combined CM cluster\n",
    "    my_subseur <- NormalizeData(my_subseur)\n",
    "    my_subseur <- ScaleData(my_subseur)\n",
    "    \n",
    "    # Run differential expression comparisons\n",
    "    aldo_vs_ctrl <- FindMarkers(my_subseur, ident.1 = \"ALDO\", ident.2 = \"CTRL\", assay = \"RNA\", test.use = \"wilcox\", logfc.threshold = 0)\n",
    "    write.csv(aldo_vs_ctrl, file = paste0(\"DEGs/nothreshold/ALDO_vs_CTRL_WRSum/degs_wilcox_\", LA_LV, \"_CM.csv\"))\n",
    "    \n",
    "    rec_vs_ctrl <- FindMarkers(my_subseur, ident.1 = \"REC\", ident.2 = \"CTRL\", assay = \"RNA\", test.use = \"wilcox\", logfc.threshold = 0)\n",
    "    write.csv(rec_vs_ctrl, file = paste0(\"DEGs/nothreshold/REC_vs_CTRL_WRSum/degs_wilcox_\", LA_LV, \"_CM.csv\"))\n",
    "    \n",
    "    aldo_vs_rec <- FindMarkers(my_subseur, ident.1 = \"ALDO\", ident.2 = \"REC\", assay = \"RNA\", test.use = \"wilcox\", logfc.threshold = 0)\n",
    "    write.csv(aldo_vs_rec, file = paste0(\"DEGs/nothreshold/ALDO_vs_REC_WRSum/degs_wilcox_\", LA_LV, \"_CM.csv\"))\n",
    "    \n",
    "    # Return results in a list\n",
    "    return(list(ALDO_vs_CTRL = aldo_vs_ctrl, REC_vs_CTRL = rec_vs_ctrl, ALDO_vs_REC = aldo_vs_rec))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab2d79b1-a062-45fc-8c58-d6a8c5b4f05d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "doCMAnalysis(\"LV\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "889cc773-4e51-4afa-9883-4dedb57bc1ed",
   "metadata": {},
   "source": [
    "### filter mt-genes, p_val_adj < 0.05 and avg_log2FC < -0.58 | avg_log2FC > 0.58"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "536a7c25-377b-4217-9b9b-ef60f07de5d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# mt_genes.csv is a list of mt-genes with >0 counts in the dataset\n",
    "\n",
    "# Load mt-genes\n",
    "mt_genes <- unlist(read.csv(\"DEGs/mt_genes.csv\", header = FALSE))\n",
    "\n",
    "# List of DEG files ALDO vs. CTRL\n",
    "file_names <- list.files(\"DEGs/ALDO_vs_CTRL_WRSum\", pattern = \"*.csv\", full.names = TRUE)\n",
    "\n",
    "# Load and process DEG files\n",
    "listDEG <- lapply(file_names, function(deg_list) {\n",
    "    read.csv(deg_list, header = TRUE, row.names = 1)\n",
    "})\n",
    "\n",
    "# Add filtering criteria\n",
    "listDEG <- lapply(listDEG, function(deg_list) {\n",
    "    deg_list <- cbind(deg_list, Gene_names = rownames(deg_list))\n",
    "    deg_list_filtered <- deg_list %>%\n",
    "        filter(\n",
    "            !Gene_names %in% mt_genes,                  # Exclude mt-genes\n",
    "            (avg_log2FC < -0.58 | avg_log2FC > 0.58),   # Filter by avg_log2FC\n",
    "            p_val_adj < 0.05                            # Filter by adjusted p-value\n",
    "        )\n",
    "    deg_list_filtered <- deg_list_filtered[, !names(deg_list_filtered) %in% c(\"Gene_names\")]\n",
    "    return(deg_list_filtered)\n",
    "})\n",
    "\n",
    "# Assign names to the DEG list\n",
    "listDEG_names <- gsub(\"\\\\..*\", \"\", gsub(\".*wilcox_\", \"\", file_names))\n",
    "names(listDEG) <- listDEG_names\n",
    "\n",
    "# Save filtered DEG files\n",
    "sapply(names(listDEG), function(name) {\n",
    "    write.csv(\n",
    "        listDEG[[name]],\n",
    "        file = paste0(\"DEGs/ALDO_vs_CTRL_WRSum/filtered/degs_wilcox_\", name, \"_filtered.csv\")\n",
    "    )\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a9b9e14-3275-4f07-850b-3971a9114d46",
   "metadata": {},
   "outputs": [],
   "source": [
    "# List of DEG files ALDO vs. REC\n",
    "file_names <- list.files(\"DEGs/ALDO_vs_REC_WRSum\", pattern = \"*.csv\", full.names = TRUE)\n",
    "\n",
    "# Load and process DEG files\n",
    "listDEG <- lapply(file_names, function(deg_list) {\n",
    "    read.csv(deg_list, header = TRUE, row.names = 1)\n",
    "})\n",
    "\n",
    "# Add filtering criteria\n",
    "listDEG <- lapply(listDEG, function(deg_list) {\n",
    "    deg_list <- cbind(deg_list, Gene_names = rownames(deg_list))\n",
    "    deg_list_filtered <- deg_list %>%\n",
    "        filter(\n",
    "            !Gene_names %in% mt_genes,                  # Exclude mt-genes\n",
    "            (avg_log2FC < -0.58 | avg_log2FC > 0.58),   # Filter by avg_log2FC\n",
    "            p_val_adj < 0.05                            # Filter by adjusted p-value\n",
    "        )\n",
    "    deg_list_filtered <- deg_list_filtered[, !names(deg_list_filtered) %in% c(\"Gene_names\")]\n",
    "    return(deg_list_filtered)\n",
    "})\n",
    "\n",
    "# Assign names to the DEG list\n",
    "listDEG_names <- gsub(\"\\\\..*\", \"\", gsub(\".*wilcox_\", \"\", file_names))\n",
    "names(listDEG) <- listDEG_names\n",
    "\n",
    "# Save filtered DEG files\n",
    "sapply(names(listDEG), function(name) {\n",
    "    write.csv(\n",
    "        listDEG[[name]],\n",
    "        file = paste0(\"DEGs/ALDO_vs_REC_WRSum/filtered/degs_wilcox_\", name, \"_filtered.csv\")\n",
    "    )\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "988ed5dc-c751-44ff-976e-71c7b309cd2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# List of DEG files REC vs. CTRL\n",
    "file_names <- list.files(\"DEGs/REC_vs_CTRL_WRSum\", pattern = \"*.csv\", full.names = TRUE)\n",
    "\n",
    "# Load and process DEG files\n",
    "listDEG <- lapply(file_names, function(deg_list) {\n",
    "    read.csv(deg_list, header = TRUE, row.names = 1)\n",
    "})\n",
    "\n",
    "# Add filtering criteria\n",
    "listDEG <- lapply(listDEG, function(deg_list) {\n",
    "    deg_list <- cbind(deg_list, Gene_names = rownames(deg_list))\n",
    "    deg_list_filtered <- deg_list %>%\n",
    "        filter(\n",
    "            !Gene_names %in% mt_genes,                  # Exclude mt-genes\n",
    "            (avg_log2FC < -0.58 | avg_log2FC > 0.58),   # Filter by avg_log2FC\n",
    "            p_val_adj < 0.05                            # Filter by adjusted p-value\n",
    "        )\n",
    "    deg_list_filtered <- deg_list_filtered[, !names(deg_list_filtered) %in% c(\"Gene_names\")]\n",
    "    return(deg_list_filtered)\n",
    "})\n",
    "\n",
    "# Assign names to the DEG list\n",
    "listDEG_names <- gsub(\"\\\\..*\", \"\", gsub(\".*wilcox_\", \"\", file_names))\n",
    "names(listDEG) <- listDEG_names\n",
    "\n",
    "# Save filtered DEG files\n",
    "sapply(names(listDEG), function(name) {\n",
    "    write.csv(\n",
    "        listDEG[[name]],\n",
    "        file = paste0(\"DEGs/REC_vs_CTRL_WRSum/filtered/degs_wilcox_\", name, \"_filtered.csv\")\n",
    "    )\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63814330-e498-49be-9c99-bf0bd00b6d22",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create REC_vs_ALDO from ALDO_vs_REC by changing sign of avg_log2FC column\n",
    "# Directory containing the original files\n",
    "input_dir <- \"DEGs/ALDO_vs_REC_WRSum/filtered/\"\n",
    "\n",
    "# Directory to save modified files\n",
    "output_dir <- \"DEGs/REC_vs_ALDO_WRSum/filtered/\"\n",
    "\n",
    "# Create output directory if it doesn't exist\n",
    "if (!dir.exists(output_dir)) {\n",
    "  dir.create(output_dir, recursive = TRUE)\n",
    "}\n",
    "\n",
    "# List all CSV files in the input directory\n",
    "file_names <- list.files(input_dir, pattern = \"\\\\.csv$\", full.names = TRUE)\n",
    "\n",
    "# Process each file\n",
    "lapply(file_names, function(file) {\n",
    "  # Read the CSV file\n",
    "  data <- read.csv(file, header = TRUE)\n",
    "  \n",
    "  # Switch the sign of the avg_log2FC column\n",
    "  if (\"avg_log2FC\" %in% colnames(data)) {\n",
    "    data$avg_log2FC <- -data$avg_log2FC\n",
    "  } else {\n",
    "    stop(paste(\"Column 'avg_log2FC' not found in file:\", file))\n",
    "  }\n",
    "  \n",
    "  # Construct the output file path\n",
    "  output_file <- file.path(output_dir, basename(file))\n",
    "  \n",
    "  # Save the modified data to the output directory\n",
    "  write.csv(data, output_file, row.names = FALSE)\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1aa7f1be-514f-4c7f-a6b1-cf6e1c34d480",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Create REC_vs_ALDO from ALDO_vs_REC by changing sign of avg_log2FC column (for unfiltered)\n",
    "input_dir <- \"DEGs/ALDO_vs_REC_WRSum/\"\n",
    "\n",
    "output_dir <- \"DEGs/REC_vs_ALDO_WRSum/\"\n",
    "\n",
    "if (!dir.exists(output_dir)) {\n",
    "  dir.create(output_dir, recursive = TRUE)\n",
    "}\n",
    "\n",
    "file_names <- list.files(input_dir, pattern = \"\\\\.csv$\", full.names = TRUE)\n",
    "\n",
    "lapply(file_names, function(file) {\n",
    "  data <- read.csv(file, header = TRUE)\n",
    "  \n",
    "  if (\"avg_log2FC\" %in% colnames(data)) {\n",
    "    data$avg_log2FC <- -data$avg_log2FC\n",
    "  } else {\n",
    "    stop(paste(\"Column 'avg_log2FC' not found in file:\", file))\n",
    "  }\n",
    "  \n",
    "  output_file <- file.path(output_dir, basename(file))\n",
    "  \n",
    "  write.csv(data, output_file, row.names = FALSE)\n",
    "})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40c512f4-090a-4638-bdf3-2593a4abcfa6",
   "metadata": {},
   "source": [
    "### Filter only protein coding genes (remove GM... and ...Rik) and non mt-genes "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a48d702-0366-42b0-93ac-5f49f578fbf2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the list of protein-coding genes\n",
    "protein_coding_genes <- unlist(read.csv(\"DEGs/nothreshold/protein_coding_gene_names_filtered.txt\", header = TRUE, stringsAsFactors = FALSE))\n",
    "\n",
    "# List of DEG files ALDO vs. CTRL\n",
    "file_names <- list.files(\"DEGs/nothreshold/ALDO_vs_CTRL_WRSum\", pattern = \"*.csv\", full.names = TRUE)\n",
    "\n",
    "# Load and process DEG files\n",
    "listDEG <- lapply(file_names, function(deg_list) {\n",
    "    read.csv(deg_list, header = TRUE, row.names = 1)\n",
    "})\n",
    "\n",
    "# Add filtering criteria\n",
    "listDEG <- lapply(listDEG, function(deg_list) {\n",
    "    deg_list <- cbind(deg_list, Gene_names = rownames(deg_list))\n",
    "    deg_list_filtered <- deg_list %>%\n",
    "        filter(\n",
    "            Gene_names %in% protein_coding_genes               # Keep only protein-coding genes\n",
    "        )\n",
    "    deg_list_filtered <- deg_list_filtered[, !names(deg_list_filtered) %in% c(\"Gene_names\")]\n",
    "    return(deg_list_filtered)\n",
    "})\n",
    "\n",
    "# Assign names to the DEG list\n",
    "listDEG_names <- gsub(\"\\\\..*\", \"\", gsub(\".*wilcox_\", \"\", file_names))\n",
    "names(listDEG) <- listDEG_names\n",
    "\n",
    "# Save filtered DEG files\n",
    "sapply(names(listDEG), function(name) {\n",
    "    write.csv(\n",
    "        listDEG[[name]],\n",
    "        file = paste0(\"DEGs/nothreshold/ALDO_vs_CTRL_WRSum/filtered/degs_wilcox_\", name, \"_filtered.csv\")\n",
    "    )\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "441ce432-b1ad-484c-b182-8c70560d7141",
   "metadata": {},
   "outputs": [],
   "source": [
    "# filter only protein coding genes (remove GM... and ...Rik) and non mt-genes \n",
    "\n",
    "# Load the list of protein-coding genes\n",
    "protein_coding_genes <- unlist(read.csv(\"DEGs/nothreshold/protein_coding_gene_names_filtered.txt\", header = TRUE, stringsAsFactors = FALSE))\n",
    "\n",
    "# List of DEG files for ALDO vs. REC\n",
    "file_names <- list.files(\"DEGs/nothreshold/ALDO_vs_REC_WRSum\", pattern = \"*.csv\", full.names = TRUE)\n",
    "\n",
    "# Load and process DEG files\n",
    "listDEG <- lapply(file_names, function(deg_list) {\n",
    "    read.csv(deg_list, header = TRUE, row.names = 1)\n",
    "})\n",
    "\n",
    "# Add filtering criteria\n",
    "listDEG <- lapply(listDEG, function(deg_list) {\n",
    "    deg_list <- cbind(deg_list, Gene_names = rownames(deg_list))\n",
    "    deg_list_filtered <- deg_list %>%\n",
    "        filter(\n",
    "            Gene_names %in% protein_coding_genes               # Keep only protein-coding genes\n",
    "        )\n",
    "    deg_list_filtered <- deg_list_filtered[, !names(deg_list_filtered) %in% c(\"Gene_names\")]\n",
    "    return(deg_list_filtered)\n",
    "})\n",
    "\n",
    "# Assign names to the DEG list\n",
    "listDEG_names <- gsub(\"\\\\..*\", \"\", gsub(\".*wilcox_\", \"\", file_names))\n",
    "names(listDEG) <- listDEG_names\n",
    "\n",
    "# Save filtered DEG files\n",
    "sapply(names(listDEG), function(name) {\n",
    "    write.csv(\n",
    "        listDEG[[name]],\n",
    "        file = paste0(\"DEGs/nothreshold/ALDO_vs_REC_WRSum/filtered/degs_wilcox_\", name, \"_filtered.csv\")\n",
    "    )\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e9a170d-d731-4b99-8152-868ca0b4c70a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# filter only protein coding genes (remove GM... and ...Rik) and non mt-genes \n",
    "\n",
    "# Load the list of protein-coding genes\n",
    "protein_coding_genes <- unlist(read.csv(\"DEGs/nothreshold/protein_coding_gene_names_filtered.txt\", header = TRUE, stringsAsFactors = FALSE))\n",
    "\n",
    "# List of DEG files for REC vs. CTRL\n",
    "file_names <- list.files(\"DEGs/nothreshold/REC_vs_CTRL_WRSum\", pattern = \"*.csv\", full.names = TRUE)\n",
    "\n",
    "# Load and process DEG files\n",
    "listDEG <- lapply(file_names, function(deg_list) {\n",
    "    read.csv(deg_list, header = TRUE, row.names = 1)\n",
    "})\n",
    "\n",
    "# Add filtering criteria\n",
    "listDEG <- lapply(listDEG, function(deg_list) {\n",
    "    deg_list <- cbind(deg_list, Gene_names = rownames(deg_list))\n",
    "    deg_list_filtered <- deg_list %>%\n",
    "        filter(\n",
    "            Gene_names %in% protein_coding_genes               # Keep only protein-coding genes\n",
    "        )\n",
    "    deg_list_filtered <- deg_list_filtered[, !names(deg_list_filtered) %in% c(\"Gene_names\")]\n",
    "    return(deg_list_filtered)\n",
    "})\n",
    "\n",
    "# Assign names to the DEG list\n",
    "listDEG_names <- gsub(\"\\\\..*\", \"\", gsub(\".*wilcox_\", \"\", file_names))\n",
    "names(listDEG) <- listDEG_names\n",
    "\n",
    "# Save filtered DEG files\n",
    "sapply(names(listDEG), function(name) {\n",
    "    write.csv(\n",
    "        listDEG[[name]],\n",
    "        file = paste0(\"DEGs/nothreshold/REC_vs_CTRL_WRSum/filtered/degs_wilcox_\", name, \"_filtered.csv\")\n",
    "    )\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1514761e-f7b9-4eeb-b3f8-6a93ad38a664",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create REC_vs_ALDO from ALDO_vs_REC by changing sign of avg_log2FC column (for unfiltered)\n",
    "input_dir <- \"DEGs/nothreshold/ALDO_vs_REC_WRSum/filtered/\"\n",
    "\n",
    "output_dir <- \"DEGs/nothreshold/REC_vs_ALDO_WRSum/filtered/\"\n",
    "\n",
    "if (!dir.exists(output_dir)) {\n",
    "  dir.create(output_dir, recursive = TRUE)\n",
    "}\n",
    "\n",
    "file_names <- list.files(input_dir, pattern = \"\\\\.csv$\", full.names = TRUE)\n",
    "\n",
    "lapply(file_names, function(file) {\n",
    "  data <- read.csv(file, header = TRUE)\n",
    "  \n",
    "  if (\"avg_log2FC\" %in% colnames(data)) {\n",
    "    data$avg_log2FC <- -data$avg_log2FC\n",
    "  } else {\n",
    "    stop(paste(\"Column 'avg_log2FC' not found in file:\", file))\n",
    "  }\n",
    "  \n",
    "  output_file <- file.path(output_dir, basename(file))\n",
    "  \n",
    "  write.csv(data, output_file, row.names = FALSE)\n",
    "})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7451cb59-73fb-4bcc-9984-b619cdb6ea3c",
   "metadata": {},
   "source": [
    "# MR target gene signature"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8cb8d1b4-8be5-4cdc-8b47-2a826d87fa76",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run for MT target genes and use normalized counts directly, no weighing by log2FC\n",
    "# Subset Seurat object by cluster\n",
    "cluster_ID <- \"CM\"\n",
    "obj_cluster <- subset(obj, subset = cell_type_CMcomb == cluster_ID)\n",
    "\n",
    "# Get sample-wise metadata\n",
    "sample_metadata <- obj_cluster@meta.data$treatment\n",
    "\n",
    "# Extract key genes\n",
    "key_genes <- c(\"Per1\", \"Per3\", \"Dbp\", \"Zbtb16\", \"Hlf\")\n",
    "\n",
    "# Get expression data for key genes from the Seurat object\n",
    "expr_data <- FetchData(obj_cluster, vars = key_genes)\n",
    "\n",
    "# Get sample IDs and add them to expression data\n",
    "expr_data$sample_id <- obj_cluster@meta.data$sample_id\n",
    "\n",
    "# Reshape the data to long format and calculate mean expression per sample\n",
    "expr_data_long <- expr_data %>%\n",
    "  gather(key = \"gene\", value = \"expression\", -sample_id) %>%\n",
    "  group_by(sample_id, gene) %>%\n",
    "  summarize(mean_expression = mean(expression), .groups = \"drop\")\n",
    "\n",
    "# Min-max scale the mean expression for each gene to be between 0 and 1\n",
    "scaled_expr_data <- expr_data_long %>%\n",
    "  group_by(gene) %>%\n",
    "  mutate(scaled_expression = scales::rescale(mean_expression)) %>%\n",
    "  ungroup()\n",
    "\n",
    "# Create scores based on the scaled expression values\n",
    "scores <- data.frame(sample_id = unique(scaled_expr_data$sample_id))\n",
    "\n",
    "# Create table with \"score_XXX\" solumns with \"XXX\" being an entry of key_genes\n",
    "for (i in 1:length(key_genes)) {\n",
    "  gene_expression <- scaled_expr_data$scaled_expression[scaled_expr_data$gene == key_genes[i]]\n",
    "  scores[paste0(\"score_\", key_genes[i])] <- gene_expression\n",
    "}\n",
    "\n",
    "# Sum scores across key genes to get total score for each sample\n",
    "scores$total_score <- rowSums(scores[, -1], na.rm = TRUE)\n",
    "\n",
    "# Normalize total score across samples\n",
    "scores$normalized_score <- scales::rescale(scores$total_score)\n",
    "\n",
    "# Add sample metadata to the weighted scores table for better visualization\n",
    "scores$treatment <- obj_cluster@meta.data$treatment[match(scores$sample_id, obj_cluster@meta.data$sample_id)]\n",
    "\n",
    "# View results\n",
    "scores\n",
    "\n",
    "# Save results as CSV\n",
    "write.csv(scores, paste0(\"DEGs/nothreshold/recovery_score/GOI_polygenic_recovery_scores_scaled_\", cluster_ID, \".csv\"), row.names = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4ec1758a-f269-45c4-a14a-d22537daf6b1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.3.3 (2024-02-29)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Ubuntu 22.04.2 LTS\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /media/daten/dmeral/micromamba/envs/scrna_dm/lib/libopenblasp-r0.3.27.so;  LAPACK version 3.12.0\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "time zone: Etc/UTC\n",
       "tzcode source: system (glibc)\n",
       "\n",
       "attached base packages:\n",
       "[1] stats4    stats     graphics  grDevices utils     datasets  methods  \n",
       "[8] base     \n",
       "\n",
       "other attached packages:\n",
       " [1] sva_3.50.0                  BiocParallel_1.36.0        \n",
       " [3] genefilter_1.84.0           mgcv_1.9-1                 \n",
       " [5] nlme_3.1-166                scales_1.3.0               \n",
       " [7] pals_1.9                    viridis_0.6.5              \n",
       " [9] viridisLite_0.4.2           svglite_2.1.3              \n",
       "[11] Rsamtools_2.18.0            Biostrings_2.70.3          \n",
       "[13] XVector_0.42.0              EnhancedVolcano_1.20.0     \n",
       "[15] ggrepel_0.9.6               edgeR_4.0.16               \n",
       "[17] limma_3.58.1                DESeq2_1.42.1              \n",
       "[19] SummarizedExperiment_1.32.0 Biobase_2.62.0             \n",
       "[21] MatrixGenerics_1.14.0       matrixStats_1.4.1          \n",
       "[23] GenomicRanges_1.54.1        GenomeInfoDb_1.38.8        \n",
       "[25] IRanges_2.36.0              S4Vectors_0.40.2           \n",
       "[27] BiocGenerics_0.48.1         SeuratDisk_0.0.0.9021      \n",
       "[29] presto_1.0.0                data.table_1.16.4          \n",
       "[31] Rcpp_1.0.13-1               lubridate_1.9.4            \n",
       "[33] forcats_1.0.0               stringr_1.5.1              \n",
       "[35] dplyr_1.1.4                 purrr_1.0.2                \n",
       "[37] readr_2.1.5                 tidyr_1.3.1                \n",
       "[39] tibble_3.2.1                tidyverse_2.0.0            \n",
       "[41] ggplot2_3.5.1               Seurat_5.1.0               \n",
       "[43] SeuratObject_5.0.2          sp_2.1-4                   \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] RcppAnnoy_0.0.22        splines_4.3.3           later_1.4.1            \n",
       "  [4] pbdZMQ_0.3-13           bitops_1.0-9            polyclip_1.10-7        \n",
       "  [7] XML_3.99-0.17           fastDummies_1.7.4       lifecycle_1.0.4        \n",
       " [10] globals_0.16.3          lattice_0.22-6          hdf5r_1.3.11           \n",
       " [13] MASS_7.3-60.0.1         magrittr_2.0.3          plotly_4.10.4          \n",
       " [16] httpuv_1.6.15           sctransform_0.4.1.9001  spam_2.11-0            \n",
       " [19] spatstat.sparse_3.1-0   reticulate_1.40.0       DBI_1.2.3              \n",
       " [22] mapproj_1.2.11          cowplot_1.1.3           pbapply_1.7-2          \n",
       " [25] RColorBrewer_1.1-3      maps_3.4.2.1            abind_1.4-8            \n",
       " [28] zlibbioc_1.48.2         Rtsne_0.17              RCurl_1.98-1.16        \n",
       " [31] GenomeInfoDbData_1.2.11 irlba_2.3.5.1           listenv_0.9.1          \n",
       " [34] spatstat.utils_3.1-1    goftest_1.2-3           RSpectra_0.16-2        \n",
       " [37] annotate_1.80.0         spatstat.random_3.3-2   fitdistrplus_1.2-1     \n",
       " [40] parallelly_1.40.1       leiden_0.4.3.1          codetools_0.2-20       \n",
       " [43] DelayedArray_0.28.0     tidyselect_1.2.1        farver_2.1.2           \n",
       " [46] base64enc_0.1-3         spatstat.explore_3.3-3  jsonlite_1.8.9         \n",
       " [49] progressr_0.15.1        ggridges_0.5.6          survival_3.7-0         \n",
       " [52] systemfonts_1.1.0       tools_4.3.3             ica_1.0-3              \n",
       " [55] glue_1.8.0              gridExtra_2.3           SparseArray_1.2.4      \n",
       " [58] IRdisplay_1.1           withr_3.0.2             fastmap_1.2.0          \n",
       " [61] fansi_1.0.6             digest_0.6.37           timechange_0.3.0       \n",
       " [64] R6_2.5.1                mime_0.12               colorspace_2.1-1       \n",
       " [67] scattermore_1.2         tensor_1.5              RSQLite_2.3.9          \n",
       " [70] dichromat_2.0-0.1       spatstat.data_3.1-4     utf8_1.2.4             \n",
       " [73] generics_0.1.3          httr_1.4.7              htmlwidgets_1.6.4      \n",
       " [76] S4Arrays_1.2.1          uwot_0.2.2              pkgconfig_2.0.3        \n",
       " [79] gtable_0.3.6            blob_1.2.4              lmtest_0.9-40          \n",
       " [82] htmltools_0.5.8.1       dotCall64_1.2           png_0.1-8              \n",
       " [85] spatstat.univar_3.1-1   tzdb_0.4.0              reshape2_1.4.4         \n",
       " [88] uuid_1.2-1              cachem_1.1.0            repr_1.1.7             \n",
       " [91] zoo_1.8-12              KernSmooth_2.23-24      parallel_4.3.3         \n",
       " [94] miniUI_0.1.1.1          AnnotationDbi_1.64.1    pillar_1.9.0           \n",
       " [97] grid_4.3.3              vctrs_0.6.5             RANN_2.6.2             \n",
       "[100] promises_1.3.2          xtable_1.8-4            cluster_2.1.7          \n",
       "[103] evaluate_1.0.1          cli_3.6.3               locfit_1.5-9.10        \n",
       "[106] compiler_4.3.3          rlang_1.1.4             crayon_1.5.3           \n",
       "[109] future.apply_1.11.3     plyr_1.8.9              stringi_1.8.4          \n",
       "[112] deldir_2.0-4            munsell_0.5.1           lazyeval_0.2.2         \n",
       "[115] spatstat.geom_3.3-4     Matrix_1.6-5            IRkernel_1.3.2         \n",
       "[118] RcppHNSW_0.6.0          hms_1.1.3               patchwork_1.3.0        \n",
       "[121] bit64_4.5.2             future_1.34.0           KEGGREST_1.42.0        \n",
       "[124] statmod_1.5.0           shiny_1.9.1             ROCR_1.0-11            \n",
       "[127] memoise_2.0.1           igraph_2.1.2            bit_4.5.0.1            "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scrna_dm",
   "language": "R",
   "name": "scrna_dm"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
